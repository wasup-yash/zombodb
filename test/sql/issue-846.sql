CREATE TABLE notnull_main (
                              pk_main SERIAL8 NOT NULL PRIMARY KEY,
                              name varchar NOT NULL
);
INSERT INTO notnull_main (name) VALUES ('Taffy');
INSERT INTO notnull_main (name) VALUES ('Tardis');
INSERT INTO notnull_main (name) VALUES ('Tackle');
INSERT INTO notnull_main (name) VALUES ('Beret');

CREATE TABLE notnull_docs (
                              pk_doc SERIAL8 NOT NULL PRIMARY KEY,
                              fk_docs_to_main int8[],
                              doctype varchar NOT NULL,
                              page_count int8
);
INSERT INTO notnull_docs (fk_docs_to_main, doctype, page_count) VALUES (ARRAY[1], 'Instructions',30);
INSERT INTO notnull_docs (fk_docs_to_main, doctype, page_count) VALUES (ARRAY[1], 'Poem',40);
INSERT INTO notnull_docs (fk_docs_to_main, doctype, page_count) VALUES (ARRAY[1,2], 'Poem',40);
INSERT INTO notnull_docs (fk_docs_to_main, doctype, page_count) VALUES (ARRAY[3], 'Instructions',50);
INSERT INTO notnull_docs (fk_docs_to_main, doctype, page_count) VALUES (ARRAY[4], 'Poem',60);

CREATE VIEW notnull_view AS
SELECT notnull_docs.pk_doc,
       notnull_docs.doctype,
       notnull_docs.page_count,
       ( SELECT array_agg(notnull_main.name) AS name
         FROM notnull_main
         WHERE notnull_main.pk_main = ANY(notnull_docs.fk_docs_to_main)
       ) AS name,
       notnull_docs.*::notnull_docs AS zdb
FROM notnull_docs;

CREATE INDEX es_idxnotnullmain
    ON notnull_main
        USING zombodb ((notnull_main.*));

CREATE INDEX es_idxnotnulldocs
    ON notnull_docs
        USING zombodb ((notnull_docs.*))
    WITH (
    max_analyze_token_count='2147483647',
    bulk_concurrency='1',
    options='fk_docs_to_main=<public.notnull_main.es_idxnotnullmain>pk_main');

SELECT * FROM notnull_view order by pk_doc;
SELECT * FROM notnull_view WHERE zdb==>'doctype=INSTRUCTIONS AND page_count <> 40' order by pk_doc;
SELECT * FROM notnull_view WHERE zdb==>'doctype=INSTRUCTIONS AND NOT page_count = 40' order by pk_doc;
SELECT upper(term) term, count
FROM zdb.tally(
        'notnull_view'::regclass,
        'name',
        'FALSE',
        '^.*',
        'doctype=INSTRUCTIONS AND page_count <> 40'::zdbquery,
        2147483647,
        'term'::termsorderby);

SELECT upper(term) term, count
FROM zdb.tally(
        'notnull_view'::regclass,
        'name',
        'FALSE',
        '^.*',
        'doctype=INSTRUCTIONS AND NOT page_count = 40'::zdbquery,
        2147483647,
        'term'::termsorderby);


drop table notnull_docs, notnull_main cascade;

