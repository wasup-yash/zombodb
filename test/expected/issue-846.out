CREATE TABLE notnull_main (
                              pk_main SERIAL8 NOT NULL PRIMARY KEY,
                              name varchar NOT NULL
);
INSERT INTO notnull_main (name) VALUES ('Taffy');
INSERT INTO notnull_main (name) VALUES ('Tardis');
INSERT INTO notnull_main (name) VALUES ('Tackle');
INSERT INTO notnull_main (name) VALUES ('Beret');
CREATE TABLE notnull_docs (
                              pk_doc SERIAL8 NOT NULL PRIMARY KEY,
                              fk_docs_to_main int8[],
                              doctype varchar NOT NULL,
                              page_count int8
);
INSERT INTO notnull_docs (fk_docs_to_main, doctype, page_count) VALUES (ARRAY[1], 'Instructions',30);
INSERT INTO notnull_docs (fk_docs_to_main, doctype, page_count) VALUES (ARRAY[1], 'Poem',40);
INSERT INTO notnull_docs (fk_docs_to_main, doctype, page_count) VALUES (ARRAY[1,2], 'Poem',40);
INSERT INTO notnull_docs (fk_docs_to_main, doctype, page_count) VALUES (ARRAY[3], 'Instructions',50);
INSERT INTO notnull_docs (fk_docs_to_main, doctype, page_count) VALUES (ARRAY[4], 'Poem',60);
CREATE VIEW notnull_view AS
SELECT notnull_docs.pk_doc,
       notnull_docs.doctype,
       notnull_docs.page_count,
       ( SELECT array_agg(notnull_main.name) AS name
         FROM notnull_main
         WHERE notnull_main.pk_main = ANY(notnull_docs.fk_docs_to_main)
       ) AS name,
       notnull_docs.*::notnull_docs AS zdb
FROM notnull_docs;
CREATE INDEX es_idxnotnullmain
    ON notnull_main
        USING zombodb ((notnull_main.*));
CREATE INDEX es_idxnotnulldocs
    ON notnull_docs
        USING zombodb ((notnull_docs.*))
    WITH (
    max_analyze_token_count='2147483647',
    bulk_concurrency='1',
    options='fk_docs_to_main=<public.notnull_main.es_idxnotnullmain>pk_main');
SELECT * FROM notnull_view order by pk_doc;
 pk_doc |   doctype    | page_count |      name      |           zdb           
--------+--------------+------------+----------------+-------------------------
      1 | Instructions |         30 | {Taffy}        | (1,{1},Instructions,30)
      2 | Poem         |         40 | {Taffy}        | (2,{1},Poem,40)
      3 | Poem         |         40 | {Taffy,Tardis} | (3,"{1,2}",Poem,40)
      4 | Instructions |         50 | {Tackle}       | (4,{3},Instructions,50)
      5 | Poem         |         60 | {Beret}        | (5,{4},Poem,60)
(5 rows)

SELECT * FROM notnull_view WHERE zdb==>'doctype=INSTRUCTIONS AND page_count <> 40' order by pk_doc;
 pk_doc |   doctype    | page_count |   name   |  zdb  
--------+--------------+------------+----------+-------
      1 | Instructions |         30 | {Taffy}  | (0,1)
      4 | Instructions |         50 | {Tackle} | (0,4)
(2 rows)

SELECT * FROM notnull_view WHERE zdb==>'doctype=INSTRUCTIONS AND NOT page_count = 40' order by pk_doc;
 pk_doc |   doctype    | page_count |   name   |  zdb  
--------+--------------+------------+----------+-------
      1 | Instructions |         30 | {Taffy}  | (0,1)
      4 | Instructions |         50 | {Tackle} | (0,4)
(2 rows)

SELECT upper(term) term, count
FROM zdb.tally(
        'notnull_view'::regclass,
        'name',
        'FALSE',
        '^.*',
        'doctype=INSTRUCTIONS AND page_count <> 40'::zdbquery,
        2147483647,
        'term'::termsorderby);
  term  | count 
--------+-------
 TACKLE |     1
 TAFFY  |     1
(2 rows)

SELECT upper(term) term, count
FROM zdb.tally(
        'notnull_view'::regclass,
        'name',
        'FALSE',
        '^.*',
        'doctype=INSTRUCTIONS AND NOT page_count = 40'::zdbquery,
        2147483647,
        'term'::termsorderby);
  term  | count 
--------+-------
 TACKLE |     1
 TAFFY  |     1
(2 rows)

drop table notnull_docs, notnull_main cascade;
